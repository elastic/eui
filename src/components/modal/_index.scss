@import '../form/variables';

@import 'modal';
