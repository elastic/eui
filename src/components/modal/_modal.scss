/**
 * 1. Fix IE overflow issue (min-height) by adding a separate wrapper for the
 *    flex display. https://github.com/philipwalton/flexbugs#flexbug-3
 * 2. IE has trouble with min-widths on flex elements. Use the pixel value
 *    from our forms since that's usually the smallest we want them.
 */

.euiModal {
  border: $euiBorderThin;
  // This mixin overwrites some of the border above
  @include euiBottomShadowLarge($adjustBorders: true); // sass-lint:disable-line mixins-before-declarations
  display: flex; /* 1 */

  position: relative;
  background-color: $euiColorEmptyShade;
  border-radius: $euiBorderRadius;
  z-index: $euiZModal;
  min-width: $euiFormMaxWidth;
  animation: euiModal $euiAnimSpeedSlow $euiAnimSlightBounce;

  .euiModal__flex { /* 1 */
    flex: 1 1 auto;
    display: flex;
    flex-direction: column;
    max-height: 75vh; // We overflow the modal body based off this
    overflow: hidden; // Ensure long, non-breaking text doesn't expand beyond the modal bounds
  }
}

.euiModal--maxWidth-default {
  max-width: map-get($euiBreakpoints, 'm');
}

.euiModal--confirmation {
  min-width: $euiFormMaxWidth;
}

.euiModalHeader {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: $euiSizeL $euiSizeXXL $euiSize $euiSizeL;
  flex-grow: 0;
  flex-shrink: 0;
}

.euiModalHeader__title {
  @include euiTitle('m');
}

.euiModalBody {
  flex-grow: 1;
  overflow: hidden;
  // The below fixes scroll on Chrome and Safari
  display: flex;
  flex-direction: column;

  .euiModalBody__overflow {
    @include euiYScrollWithShadows;
    padding: $euiSizeS $euiSizeL;
  }
}

.euiModalFooter {
  display: flex;
  justify-content: flex-end;
  padding: $euiSize $euiSizeL $euiSizeL;
  flex-grow: 0;
  flex-shrink: 0; // ensure the height of the footer is based off it's contents and doesn't squish

  > * + * {
    margin-left: $euiSize;
  }
}

// If a body doesn't exist, remove some extra padding from footer
.euiModalHeader + .euiModalFooter {
  padding-top: $euiSizeS;
}

// If a footer doesn't exist (body is the last element) add padding to the bottom
.euiModalBody:last-of-type .euiModalBody__overflow {
  padding-bottom: $euiSizeL;
}


// The actual size of the X button in pixels is a bit fuzzy because of all the
// button padding so there is some pixel pushing here.
.euiModal__closeIcon {
  background-color: transparentize($euiColorEmptyShade, .1);
  position: absolute;
  right: $euiSizeL - 7px;
  top: $euiSizeL - 7px;
  z-index: 3;
}

@keyframes euiModal {
  0% {
    opacity: 0;
    transform: translateY($euiSizeXL);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

// On mobile we fix modals as a takeover.
@include euiBreakpoint('xs', 's') {
  .euiModal {
    // sass-lint:disable-block no-important
    position: fixed;
    width: 100vw !important;
    max-width: none !important;
    min-width: 0 !important;
    left: 0;
    right: 0;
    bottom: 0;
    top: 0;
    border-radius: 0;
    border: none;

    &.euiModal--confirmation {
      @include euiBottomShadowLarge($euiShadowColorLarge, .1, false, true);
      top: auto;
    }

    .euiModal__flex { /* 1 */
      max-height: 100vh;
    }
  }

  .euiModalHeader {
    width: 100vw;
  }

  .euiModalFooter {
    background: $euiColorLightestShade;
    padding: $euiSizeM $euiSizeL !important; // sass-lint:disable-line no-important
    width: 100vw;
    justify-content: stretch;

    > * {
      flex: 1;

      + * {
        margin-left: 0;
      }
    }
  }

  .euiModalBody {
    width: 100vw;

    .euiModalBody__overflow {
      padding-bottom: $euiSizeL;
    }
  }
}
