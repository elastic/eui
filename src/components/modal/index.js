export {
  EuiConfirmModal,
  CONFIRM_BUTTON as EUI_MODAL_CONFIRM_BUTTON,
  CANCEL_BUTTON as EUI_MODAL_CANCEL_BUTTON,
} from './confirm_modal';
export { EuiModal } from './modal';
export { EuiModalFooter } from './modal_footer';
export { EuiModalHeader } from './modal_header';
export { EuiModalBody } from './modal_body';
export { EuiModalHeaderTitle } from './modal_header_title';
