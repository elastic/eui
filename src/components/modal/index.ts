/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

export {
  EuiConfirmModal,
  EuiConfirmModalProps,
  CONFIRM_BUTTON as EUI_MODAL_CONFIRM_BUTTON,
  CANCEL_BUTTON as EUI_MODAL_CANCEL_BUTTON,
} from './confirm_modal';
export { EuiModal, EuiModalProps } from './modal';
export { EuiModalFooter, EuiModalFooterProps } from './modal_footer';
export { EuiModalHeader, EuiModalHeaderProps } from './modal_header';
export { EuiModalBody, EuiModalBodyProps } from './modal_body';
export {
  EuiModalHeaderTitle,
  EuiModalHeaderTitleProps,
} from './modal_header_title';
