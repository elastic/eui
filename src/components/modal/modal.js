import React, {
  Component,
} from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import FocusTrap from 'focus-trap-react';

import { keyCodes } from '../../services';

import { EuiButtonIcon } from '../button'

export class EuiModal extends Component {
  onKeyDown = event => {
    if (event.keyCode === keyCodes.ESCAPE) {
      event.preventDefault();
      event.stopPropagation();
      this.props.onClose();
    }
  };

  render() {
    const {
      className,
      children,
      onClose, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;

    const classes = classnames('euiModal', className);

    return (
      <FocusTrap
        focusTrapOptions={{
          fallbackFocus: () => this.modal,
        }}
      >
        {
          // Create a child div instead of applying these props directly to FocusTrap, or else
          // fallbackFocus won't work.
        }
        <div
          ref={node => { this.modal = node; }}
          className={classes}
          onKeyDown={this.onKeyDown}
          tabIndex={0}
          {...rest}
        >
          <EuiButtonIcon
            iconType="cross"
            onClick={onClose}
            className="euiModal__closeIcon"
            color="text"
            aria-label="Closes this modal window"
          />
          {children}
        </div>
      </FocusTrap>
    );
  }
}

EuiModal.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
  onClose: PropTypes.func.isRequired,
};
