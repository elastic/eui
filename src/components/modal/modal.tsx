/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { FunctionComponent, ReactNode, HTMLAttributes } from 'react';
import classnames from 'classnames';

import { keys } from '../../services';

import { EuiButtonIcon } from '../button';

import { EuiFocusTrap } from '../focus_trap';
import { EuiOverlayMask } from '../overlay_mask';
import { EuiI18n } from '../i18n';

export interface EuiModalProps extends HTMLAttributes<HTMLDivElement> {
  className?: string;
  /**
   * ReactNode to render as this component's content
   */
  children: ReactNode;
  onClose: (
    event?:
      | React.KeyboardEvent<HTMLDivElement>
      | React.MouseEvent<HTMLButtonElement>
  ) => void;
  /**
   * Sets the max-width of the modal.
   * Set to `true` to use the default (`euiBreakpoints 'm'`),
   * set to `false` to not restrict the width,
   * set to a number for a custom width in px,
   * set to a string for a custom width in custom measurement.
   */
  maxWidth?: boolean | number | string;
  /**
   * Specifies what element should initially have focus.
   * Can be a DOM node, or a selector string (which will be passed to document.querySelector() to find the DOM node), or a function that returns a DOM node.
   */
  initialFocus?: HTMLElement | (() => HTMLElement) | string;
}

export const EuiModal: FunctionComponent<EuiModalProps> = ({
  className,
  children,
  initialFocus,
  onClose,
  maxWidth = true,
  style,
  ...rest
}) => {
  const onKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    if (event.key === keys.ESCAPE) {
      event.preventDefault();
      event.stopPropagation();
      onClose(event);
    }
  };

  let newStyle;
  let widthClassName;
  if (maxWidth === true) {
    widthClassName = 'euiModal--maxWidth-default';
  } else if (maxWidth !== false) {
    const value = typeof maxWidth === 'number' ? `${maxWidth}px` : maxWidth;
    newStyle = { ...style, maxWidth: value };
  }

  const classes = classnames('euiModal', widthClassName, className);

  return (
    <EuiOverlayMask>
      <EuiFocusTrap initialFocus={initialFocus}>
        {
          // Create a child div instead of applying these props directly to FocusTrap, or else
          // fallbackFocus won't work.
        }
        <div
          className={classes}
          onKeyDown={onKeyDown}
          tabIndex={0}
          style={newStyle || style}
          {...rest}>
          <EuiI18n
            token="euiModal.closeModal"
            default="Closes this modal window">
            {(closeModal: string) => (
              <EuiButtonIcon
                iconType="cross"
                onClick={onClose}
                className="euiModal__closeIcon"
                color="text"
                aria-label={closeModal}
              />
            )}
          </EuiI18n>
          <div className="euiModal__flex">{children}</div>
        </div>
      </EuiFocusTrap>
    </EuiOverlayMask>
  );
};
