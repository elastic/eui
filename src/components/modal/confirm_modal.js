import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { EuiModal } from './modal';
import { EuiModalFooter } from './modal_footer';
import { EuiModalHeader } from './modal_header';
import { EuiModalHeaderTitle } from './modal_header_title';
import { EuiModalBody } from './modal_body';

import {
  EuiButton,
  EuiButtonEmpty,
} from '../button';

import {
  EuiText,
} from '../text';

export const CONFIRM_BUTTON = 'confirm';
export const CANCEL_BUTTON = 'cancel';

const CONFIRM_MODAL_BUTTONS = [
  CONFIRM_BUTTON,
  CANCEL_BUTTON,
];

export class EuiConfirmModal extends Component {
  componentDidMount() {
    // We have to do this instead of using `autoFocus` because React's polyfill for auto-focusing
    // elements conflicts with the focus-trap logic we have on EuiModal.
    const { defaultFocusedButton } = this.props;

    // Wait a beat for the focus-trap to complete, and then set focus to the right button. Check that
    // the buttons exist first, because it's possible the modal has been closed already.
    requestAnimationFrame(() => {
      if (defaultFocusedButton === CANCEL_BUTTON && this.cancelButton) {
        this.cancelButton.focus();
      } else if (defaultFocusedButton === CONFIRM_BUTTON && this.confirmButton) {
        this.confirmButton.focus();
      }
    });
  }

  confirmRef = node => this.confirmButton = node;
  cancelRef = node => this.cancelButton = node;

  render() {
    const {
      children,
      title,
      onCancel,
      onConfirm,
      cancelButtonText,
      confirmButtonText,
      className,
      defaultFocusedButton, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;

    const classes = classnames('euiModal--confirmation', className);

    let modalTitle;

    if (title) {
      modalTitle = (
        <EuiModalHeader>
          <EuiModalHeaderTitle data-test-subj="confirmModalTitleText">
            {title}
          </EuiModalHeaderTitle>
        </EuiModalHeader>
      );
    }

    let message;

    if (typeof children === 'string') {
      message = <p>{children}</p>;
    } else {
      message = children;
    }

    return (
      <EuiModal
        className={classes}
        onClose={onCancel}
        {...rest}
      >
        {modalTitle}

        <EuiModalBody>
          <EuiText data-test-subj="confirmModalBodyText">
            {message}
          </EuiText>
        </EuiModalBody>

        <EuiModalFooter>
          <EuiButtonEmpty
            data-test-subj="confirmModalCancelButton"
            onClick={onCancel}
            size="s"
            buttonRef={this.cancelRef}
          >
            {cancelButtonText}
          </EuiButtonEmpty>

          <EuiButton
            data-test-subj="confirmModalConfirmButton"
            onClick={onConfirm}
            size="s"
            fill
            buttonRef={this.confirmRef}
          >
            {confirmButtonText}
          </EuiButton>
        </EuiModalFooter>
      </EuiModal>
    );
  }
}

EuiConfirmModal.propTypes = {
  children: PropTypes.node,
  title: PropTypes.node,
  cancelButtonText: PropTypes.node,
  confirmButtonText: PropTypes.node,
  onCancel: PropTypes.func,
  onConfirm: PropTypes.func,
  className: PropTypes.string,
  defaultFocusedButton: PropTypes.oneOf(CONFIRM_MODAL_BUTTONS)
};
