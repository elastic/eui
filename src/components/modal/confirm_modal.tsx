/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, {
  FunctionComponent,
  ReactNode,
  useEffect,
  useState,
} from 'react';
import classnames from 'classnames';

import { EuiModal, EuiModalProps } from './modal';
import { EuiModalFooter } from './modal_footer';
import { EuiModalHeader } from './modal_header';
import { EuiModalHeaderTitle } from './modal_header_title';
import { EuiModalBody } from './modal_body';

import { ButtonColor, EuiButton, EuiButtonEmpty } from '../button';

import { EuiText } from '../text';

export interface EuiConfirmModalProps
  extends Omit<
    EuiModalProps,
    'children' | 'initialFocus' | 'onClose' | 'title'
  > {
  /**
   * ReactNode to render as this component's content
   */
  children?: ReactNode;
  title?: ReactNode;
  cancelButtonText?: ReactNode;
  confirmButtonText?: ReactNode;
  onCancel: (
    event?:
      | React.KeyboardEvent<HTMLDivElement>
      | React.MouseEvent<HTMLButtonElement>
  ) => void;
  onConfirm?: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  confirmButtonDisabled?: boolean;
  className?: string;
  defaultFocusedButton?: typeof CONFIRM_BUTTON | typeof CANCEL_BUTTON;
  buttonColor?: ButtonColor;
  // For docs only, will get passed with ...rest
  /**
   * Sets the max-width of the modal.
   * Set to `true` to use the default (`euiBreakpoints 'm'`),
   * set to `false` to not restrict the width,
   * set to a number for a custom width in px,
   * set to a string for a custom width in custom measurement.
   */
  maxWidth?: boolean | number | string;
  /**
   * Passes `isLoading` prop to the confirm button
   */
  isLoading?: boolean;
}

export const CONFIRM_BUTTON = 'confirm';
export const CANCEL_BUTTON = 'cancel';

export const EuiConfirmModal: FunctionComponent<EuiConfirmModalProps> = ({
  children,
  title,
  onCancel,
  onConfirm,
  cancelButtonText,
  confirmButtonText,
  confirmButtonDisabled,
  className,
  buttonColor = 'primary',
  defaultFocusedButton,
  isLoading,
  ...rest
}) => {
  const [cancelButton, setCancelButton] = useState<
    HTMLButtonElement | HTMLAnchorElement | null
  >(null);
  const [confirmButton, setConfirmButton] = useState<HTMLButtonElement | null>(
    null
  );

  useEffect(() => {
    // We have to do this instead of using `autoFocus` because React's polyfill for auto-focusing
    // elements conflicts with the focus-trap logic we have on EuiModal.
    // Wait a beat for the focus-trap to complete, and then set focus to the right button. Check that
    // the buttons exist first, because it's possible the modal has been closed already.
    requestAnimationFrame(() => {
      if (defaultFocusedButton === CANCEL_BUTTON && cancelButton) {
        cancelButton.focus();
      } else if (defaultFocusedButton === CONFIRM_BUTTON && confirmButton) {
        confirmButton.focus();
      }
    });
  });

  const confirmRef = (node: HTMLButtonElement | null) => setConfirmButton(node);
  const cancelRef = (node: HTMLButtonElement | HTMLAnchorElement | null) =>
    setCancelButton(node);

  const classes = classnames('euiModal--confirmation', className);

  let modalTitle;

  if (title) {
    modalTitle = (
      <EuiModalHeader>
        <EuiModalHeaderTitle data-test-subj="confirmModalTitleText">
          {title}
        </EuiModalHeaderTitle>
      </EuiModalHeader>
    );
  }

  let message;

  if (typeof children === 'string' && children.length > 0) {
    message = <p>{children}</p>;
  } else {
    message = children;
  }

  return (
    <EuiModal className={classes} onClose={onCancel} {...rest}>
      {modalTitle}

      {message && (
        <EuiModalBody>
          <EuiText data-test-subj="confirmModalBodyText">{message}</EuiText>
        </EuiModalBody>
      )}

      <EuiModalFooter>
        <EuiButtonEmpty
          data-test-subj="confirmModalCancelButton"
          onClick={onCancel}
          buttonRef={cancelRef}>
          {cancelButtonText}
        </EuiButtonEmpty>

        <EuiButton
          data-test-subj="confirmModalConfirmButton"
          onClick={onConfirm}
          isLoading={isLoading}
          fill
          buttonRef={confirmRef}
          color={buttonColor}
          isDisabled={confirmButtonDisabled}>
          {confirmButtonText}
        </EuiButton>
      </EuiModalFooter>
    </EuiModal>
  );
};
