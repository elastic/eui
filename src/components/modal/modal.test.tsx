/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { mount } from 'enzyme';
import { requiredProps, takeMountedSnapshot } from '../../test';

import { EuiModal } from './modal';

test('renders EuiModal', () => {
  const component = (
    <EuiModal onClose={() => {}} {...requiredProps}>
      children
    </EuiModal>
  );

  expect(
    takeMountedSnapshot(mount(component), { hasArrayOutput: true })
  ).toMatchSnapshot();
});
