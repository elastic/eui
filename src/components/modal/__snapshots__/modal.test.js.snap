// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders EuiModal 1`] = `
<div>
  <div
    aria-label="aria-label"
    class="euiModal testClass1 testClass2"
    data-test-subj="test subject string"
    tabindex="0"
  >
    <button
      aria-label="Closes this modal window"
      class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="euiIcon euiIcon--medium euiButtonIcon__icon"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xlink="http://www.w3.org/1999/xlink"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <path
            d="M7.293 8l-4.147 4.146a.5.5 0 0 0 .708.708L8 8.707l4.146 4.147a.5.5 0 0 0 .708-.708L8.707 8l4.147-4.146a.5.5 0 0 0-.708-.708L8 7.293 3.854 3.146a.5.5 0 1 0-.708.708L7.293 8z"
            id="cross-a"
          />
        </defs>
        <use
          fill-rule="nonzero"
          href="#cross-a"
        />
      </svg>
    </button>
    children
  </div>
</div>
`;
